### PR TITLE
bpo-27535: Optimize warnings.warn()

### PR DESCRIPTION
**Pull request by vstinner**
* Optimize warnings.filterwarnings(). Replace re.compile('') with
  None to avoid the cost of calling a regex.match() method, whereas
  it always matchs.
* Optimize get_warnings_attr(): replace PyObject_GetAttrString() with
  _PyObject_GetAttrId().

Cleanup also create_filter():

* Use _Py_IDENTIFIER() to allow to cleanup strings at Python
  finalization
* Replace Py_FatalError() with a regular exceptions

https://bugs.python.org/issue27535

### CODE REVIEW

File: Python/_warnings.c [hunk 1]
```diff
@@ -1146,50 +1151,36 @@ static PyMethodDef warnings_functions[] = {
...
-            always_str = PyUnicode_InternFromString("always");
-            if (always_str == NULL)
-                return NULL;
-        }
-        action_obj = always_str;
+        id = &PyId_always;
     }
     else {
-        Py_FatalError("unknown action");
+        PyErr_SetString(PyExc_ValueError, "unknown action");
```
Comment (serhiy-storchaka) [inline Python/_warnings.c:1173]: Please revert this change. This is an internal function called with predefined set of arguments. The code in the `else` branch is a dead code. Here should be `Py_FatalError()` or `Py_UNREACHABLE()`.

File: Python/_warnings.c [hunk 2]
```diff
@@ -1146,50 +1151,36 @@ static PyMethodDef warnings_functions[] = {
 static PyObject *
 create_filter(PyObject *category, const char *action)
 {
-    static PyObject *ignore_str = NULL;
-    static PyObject *error_str = NULL;
-    static PyObject *default_str = NULL;
-    static PyObject *always_str = NULL;
-    PyObject *action_obj = NULL;
+    _Py_IDENTIFIER(ignore);
```
Comment (serhiy-storchaka) [inline Python/_warnings.c:1154]: This optimization doesn't make sense since `create_filter()` is used only few times at start. It would be better to pass `_Py_Identifier *` instead of `const char *`, this could save few lines of code.

Comment (vstinner) [reply Python/_warnings.c:1154]: Sorry, I chose to put two changes into the same commit. This change was made to allow to free strings at Python finallization, it's not an optimization.

File: Python/_warnings.c (hunk 1 above)
Comment (vstinner) [reply Python/_warnings.c:1173]: I like your idea of using _Py_Identifier: I proposed https://github.com/python/cpython/pull/4516 to implement your idea. It avoids the need to have to unhandle "unknown action".

FYI I'm trying to remove all Py_FatalError() calls from the Python source code, one by one, when it's possible and when it makes sense. I added a new _PyInitError API to report errors at Python initialization, so the caller can decide when Py_FatalError() will be called, and maybe release memory before calling Py_FatalError() (or flush files, or something else).
